taskqueue/taskqueue.go

```diff
@@ -118,14 +116,16 @@ func (tq *WorkerTaskQueue) worker(executor Executor) {
 			}
 		}
 		for _, task := range tasks {
-			atomic.AddInt32(&tq.activeTasks, 1)
+			tq.noTaskCond.L.Lock()
+			tq.activeTasks = tq.activeTasks + 1
+			tq.noTaskCond.L.Unlock()
 			terminate := executor.ExecuteTask(tq.ctx, pid, task)
-			if atomic.AddInt32(&tq.activeTasks, -1) == 0 {
-				select {
-				case tq.noTaskSignal <- struct{}{}:
-				default:
-				}
+			tq.noTaskCond.L.Lock()
+			tq.activeTasks = tq.activeTasks - 1
+			if tq.activeTasks == 0 {
+				tq.noTaskCond.Broadcast()
 			}
+			tq.noTaskCond.L.Unlock()
```

Comment:
sync.Cond.Broadcast doesn't require holding the lock to be called, so I wonder if you could make this loop lock-free by also continuing to use sync/atomic for activeTasks.

I think that's ideally what we want, because otherwise we're adding some amount of lock contention on each worker. The locks are only held for tiny amounts of time here, but they still add their overhead, whereas atomics are comparatively very cheap.

In other words, I _think_ we can just use sync.Cond for its broadcast feature, and continue using atomics for the counter. If our condition was more complex and we couldn't use atomics, then we'd need the shared lock for sure, but I don't think we absolutely need it here.

Reply:
Careful: if a `sync.Cond.Broadcast()` is issued and there isn't anyone `Wait()`ing to "hear it": the broadcast is **discarded**. You effectively need at least an implicit ( by some other means ) "lock", to ensure that there is something in `Wait()`ing

Reply:
I guess this depends on when the Wait call happens. I assume we call Wait before the workers all go idle. If we may call Wait at a later time, then indeed we need something more. Perhaps Wait should first atomically check if we're *already* idle, and if so, return immediately. Otherwise, block until the next broadcast.

Reply:
Wait, the current code already does this - it only Waits inside the loop after it checks that we're not already idle.

I think you're saying that replacing the lock with atomics on the broadcast side could add a race, like the following sequence of events:

1) WaitForNoActiveTasks grabs the lock
2) WaitForNoActiveTasks sees the counter is 1, so it enters the loop
3) worker finishes a task, and broadcasts without grabbing the lock
4) WaitForNoActiveTasks calls Cond.Wait, which doesn't see the broadcast as it is late

Whereas with Rod's current code:

1) WaitForNoActiveTasks grabs the lock
2) WaitForNoActiveTasks sees the counter is 1, so it enters the loop
3) worker finishes a task, and tries to grab the lock to broadcast - blocking, as WaitForNoActiveTasks has the lock
4) WaitForNoActiveTasks calls Cond.Wait, releasing the block temporarily
5) worker grabs the lock and broadcasts
6) WaitForNoActiveTasks sees the broadcast and finishes

So I think you're right. we don't *need* to hold the lock to broadcast, like the docs say, but not doing so inserts a form of logic race.

Reply:
Yup! This is what I was trying to convey, sorry for being terse!